File: src/screenshot.py
import time
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.keys import Keys
from PIL import Image

def take_screenshot(url, output_path, zoom_count, crop_settings):
    chrome_driver_path = "./chromedriver"
    chrome_options = webdriver.ChromeOptions()
    chrome_options.headless = True
    service = Service(executable_path=chrome_driver_path)
    
    driver = webdriver.Chrome(service=service, options=chrome_options)
    driver.set_window_size(1920, 1080)

    driver.get(url)
    time.sleep(10)

    for _ in range(zoom_count):
        driver.find_element_by_tag_name("body").send_keys(Keys.ADD)
        time.sleep(1)

    driver.save_screenshot("temp_screenshot.png")
    driver.quit()

    crop_image("temp_screenshot.png", output_path, crop_settings)

def crop_image(input_path, output_path, crop_settings):
    left, top, right, bottom = crop_settings["left"], crop_settings["top"], crop_settings["right"], crop_settings["bottom"]

    with Image.open(input_path) as img:
        width, height = img.size
        cropped_img = img.crop((left, top, width - right, height - bottom))
        cropped_img.save(output_path)

url = "https://globe.adsbexchange.com/?icao=a835af"
output_path = "screenshot.png"
zoom_count = 10
crop_settings = {"left": 250, "top": 0, "right": 650, "bottom": 0}

def main():
    start_time = time.time()

    take_screenshot(url, output_path, zoom_count, crop_settings)

    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Execution time: {execution_time:.2f} seconds")

if __name__ == "__main__":
    main()

File: src/screenshot.js
const puppeteer = require('puppeteer');
const sharp = require('sharp');
const fs = require('fs');

const ICAO_ID = fs.readFileSync('../config/ICAO-ID.txt', 'utf8').trim();

const closeZoom = 9;
const farZoom = 3;

const cropSettings = {
  left: 400,
  top: 300,
  right: 850,
  bottom: 310,
};

const url = `https://globe.adsbexchange.com/?icao=${ICAO_ID}`;

console.log(`==> Let's take screenshots`);

async function takeScreenshots(url, closeOutputPath, farOutputPath, closeZoom, farZoom, cropSettings) {
  // Launch the browser
  const browser = await puppeteer.launch({
    headless: true,
    executablePath: '/usr/bin/google-chrome', // or '/usr/bin/chromium-browser' if you are using Chromium
  });
  const page = await browser.newPage();
  await page.setViewport({ width: 1920, height: 1080 });

  // Open the URL
  console.log(`==> Open the URL`);
  await page.goto(url, { waitUntil: 'networkidle0' });

  // Wait for the map to load
  console.log(`==> Wait for the map to load`);
  await page.waitForTimeout(5000);

  // Zoom in on the map and take a close-up screenshot
  console.log(`==> Zooming in and taking a close-up screenshot`);
  for (let i = 0; i < closeZoom; i++) {
    await page.keyboard.press('+');
    await page.waitForTimeout(2000);
  }
  const closeScreenshotBuffer = await page.screenshot();
  const closeCroppedBuffer = await cropImage(closeScreenshotBuffer, cropSettings);
  await sharp(closeCroppedBuffer).toFile(closeOutputPath);

  // Zoom out and take a far-away screenshot
  console.log(`==> Zooming out and taking a far-away screenshot`);
  for (let i = 0; i < closeZoom + farZoom; i++) {
    await page.keyboard.press('-');
    await page.waitForTimeout(2000);
  }
  const farScreenshotBuffer = await page.screenshot();
  const farCroppedBuffer = await cropImage(farScreenshotBuffer, cropSettings);
  await sharp(farCroppedBuffer).toFile(farOutputPath);

  // Close the browser
  console.log(`==> Close the browser`);
  await browser.close();
}

async function cropImage(imageBuffer, cropSettings) {
  const { left, top, right, bottom } = cropSettings;
  const originalImage = sharp(imageBuffer);
  const { width, height } = await originalImage.metadata();

  const croppedWidth = width - left - right;
  const croppedHeight = height - top - bottom;

  return originalImage.extract({ left, top, width: croppedWidth, height: croppedHeight }).toBuffer();
}

(async () => {
  const startTime = Date.now();

  await takeScreenshots(url, 'close.png', 'far.png', closeZoom, farZoom, cropSettings);

  const endTime = Date.now();
  const executionTime = (endTime - startTime) / 1000;
  console.log(`==> Execution time: ${executionTime.toFixed(2)} seconds`);
})();


